agb/src/interrupt.rs

```diff
@@ -21,37 +28,344 @@ pub enum Interrupt {
+                unsafe { &*c }.next.set(self.next.get());
+                return;
+            }
+            c = p;
+        }
+    }
+}
+
+fn interrupt_to_root(interrupt: Interrupt) -> &'static InterruptRoot {
+    unsafe { &INTERRUPT_TABLE[interrupt as usize] }
+}
+
+fn get_interrupt_handle_root<'a>(
+    f: &'a dyn Fn(),
+    root: &InterruptRoot,
+) -> InterruptClosureBounded<'a> {
+    InterruptClosureBounded {
+        c: InterruptClosure {
+            closure: unsafe { core::mem::transmute(f as *const _) },
+            next: Cell::new(core::ptr::null()),
+            root: root as *const _,
+        },
+        _phantom: PhantomData,
+        _unpin: PhantomPinned,
+    }
+}
+
+pub fn get_interrupt_handle(
+    f: &(dyn Fn() + Send + Sync),
+    interrupt: Interrupt,
+) -> InterruptClosureBounded {
+    let root = interrupt_to_root(interrupt);
+
+    get_interrupt_handle_root(f, root)
+}
+
+pub fn add_interrupt<'a>(interrupt: Pin<&'a InterruptClosureBounded<'a>>) {
+    let root = unsafe { &*interrupt.c.root };
+    root.add();
+    let mut c = root.next.get();
+    if c.is_null() {
+        root.next.set((&interrupt.c) as *const _);
+        return;
+    }
+    loop {
+        let p = unsafe { &*c }.next.get();
+        if p.is_null() {
+            unsafe { &*c }.next.set((&interrupt.c) as *const _);
+            return;
+        }
+
+        c = p;
+    }
+}
+
+#[macro_export]
+macro_rules! add_interrupt_handler {
+    ($interrupt: expr, $handler: expr) => {
+        let a = $handler;
+        let a = $crate::interrupt::get_interrupt_handle(&a, $interrupt);
+        let a = unsafe { core::pin::Pin::new_unchecked(&a) };
+        $crate::interrupt::add_interrupt(a);
+    };
+}
+
+#[test_case]
+fn test_vblank_interrupt_handler(_gba: &mut crate::Gba) {
+    {
+        let counter = Mutex::new(0);
+        let counter_2 = Mutex::new(0);
+        add_interrupt_handler!(Interrupt::VBlank, || *counter.lock() += 1);
+        add_interrupt_handler!(Interrupt::VBlank, || *counter_2.lock() += 1);
+
+        let vblank = VBlank::get();
+
+        while *counter.lock() < 100 || *counter_2.lock() < 100 {
+            vblank.wait_for_vblank();
+        }
+    }
+
+    assert_eq!(
+        interrupt_to_root(Interrupt::VBlank).next.get(),
+        core::ptr::null(),
+        "expected the interrupt table for vblank to be empty"
+    );
+}
+
+#[test_case]
+fn test_interrupt_table_length(_gba: &mut crate::Gba) {
+    assert_eq!(
+        unsafe { INTERRUPT_TABLE.len() },
+        Interrupt::Gamepak as usize + 1,
+        "interrupt table should be able to store gamepak interrupt"
+    );
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+enum MutexState {
+    Unlocked,
+    Locked(bool),
+}
+
+pub struct Mutex<T> {
+    internal: UnsafeCell<T>,
+    state: UnsafeCell<MutexState>,
+}
+
+unsafe impl<T> Send for Mutex<T> {}
+unsafe impl<T> Sync for Mutex<T> {}
+
+impl<T> Mutex<T> {
+    pub fn lock(&self) -> MutexRef<T> {
```

Comment:
the real rust implementation of MutexRef takes a reference to the underlying mutex and also has a restricted lifetime the same as the mutex here.

This is unsafe I believe because the mutex can go out of scope, causing what it is holding to be dropped but you can keep a hold of the ref here and still be able to access the memory it is holding.

Reply:
Taking the underlying mutex in the mutex ref feels like it would simplify the implementation of drop on the mutex ref too

Reply:
If you could drop the Mutex before dropping the Ref then you've managed to prove the borrow checker is wrong, as this would leave an invalid reference due to lifetimes.

Reply:
```rust
let foo = {
  let something = Something::new();
  let bar = Mutex::new(something);
  bar.lock();
};
```

This drops something and the mutex but you're still holding on to the MutexRef?

Reply:
Well, it doesn't because foo is (), (that expression doesn't evaluate to anything with the semicolon there).
In practice, this happens:
![image](https://user-images.githubusercontent.com/8143879/127922170-92410ece-aaa0-4440-8689-467aa87f0e79.png)

Reply:
I clearly don't understand the borrow checker well enough to write unsafe code :joy:

Reply:
Maybe it's because I've elided some lifetimes that you're confused.
If you look at the definition of MutexRef it is
```rust
pub struct MutexRef<'a, T>
```
It takes a lifetime `'a`. Now we look back at the lock function
```rust
pub fn lock(&self) -> MutexRef<T>
```
No lifetimes! But this is because of lifetime elision, so it is actually
```rust
pub fn lock(&'a self) -> MutexRef<'a, T>
```
which means that the borrow checker will enforce that MutexRef lasts no longer than `'a`, and therefore Mutex has to exist the entire time we have MutexRef. This is the only borrow checker rule we care about for this (as the other rules are to do with mut references).